### PR TITLE
Feature: Stash Warning Compact

### PR DESCRIPTION
**Pull request by DavidArthurCole**
## What
Added a compactor for the stash warning messages.

## Changelog New Features
+ Added chat compacting for 'items in stash' warnings. - Daveed

### CODE REVIEW

**Comment by DavidArthurCole** [comment]
Fire, this already got done in a similar way in #2013, which has had conflicts since June. So, I guess this is a continuation PR, now?

**Comment by hannibal002** [comment]
> Fire, this already got done in a similar way in #2013, which has had conflicts since June. So, I guess this is a continuation PR, now?

agreed. will close the other pr

File: src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                "§8totalling $lastDifferingMaterialsCount ${StringUtils.pluralize(lastDifferingMaterialsCount, "type")}§7. " +
+                "§3Click to pickup§7.",
+            onClick = {
+                HypixelCommands.pickupStash()
+            },
+            prefix = false,
+        )
+    }
+
+    private fun isEnabled() = config.stashWarnings != ChatConfig.StashHandlerType.NONE
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt]: missing skyblock check

File: src/main/java/at/hannibal2/skyhanni/config/features/chat/ChatConfig.java [hunk 2]
```diff
@@ -142,4 +143,27 @@ public String toString() {
...
     @FeatureToggle
     public boolean petRarityDropMessage = true;
+
+    @Expose
+    @ConfigOption(name = "Stash Warnings", desc = "Compact or hide warnings relating to items/materials in your stash.")
+    @ConfigEditorDropdown
+    public StashHandlerType stashWarnings = StashHandlerType.COMPACT;
+
+    public enum StashHandlerType {
+        NONE("None (Show all)"),
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/chat/ChatConfig.java]: Maybe could indicate somehow that this means showing the default hypixel stash message

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
Some things I saw
would also be good to include an image of the option both on compact mode and the default hypixel message but these arent necessary

**Review by CalMWolfs — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (!isEnabled()) return
+
+        genericAddedToStashPattern.matchMatcher(event.message) {
+            event.blockedReason = "stash_compact"
+        }
+
+        materialCountPattern.matchMatcher(event.message) {
+            groupOrNull("count")?.replace(",", "")?.toIntOrNull()?.let { count ->
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt]: `.replace(",", "")?.toIntOrNull()` can be replaced with `.formatIntOrNull()`

File: src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                lastMaterialCount = count
+            }
+            event.blockedReason = "stash_compact"
+        }
+
+        differingMaterialsCountPattern.matchMatcher(event.message) {
+            groupOrNull("count")?.replace(",", "")?.toIntOrNull()?.let { count ->
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt]: `.replace(",", "")?.toIntOrNull()` can be replaced with `.formatIntOrNull()`

File: src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    private fun sendCompactedStashMessage() {
+        ChatUtils.clickableChat(
+            "§7You have §3${lastMaterialCount} §7${StringUtils.pluralize(lastMaterialCount, lastType)} in stash, " +
+                "§8totalling $lastDifferingMaterialsCount ${StringUtils.pluralize(lastDifferingMaterialsCount, "type")}§7. " +
+                "§3Click to ${if(config.useViewStash) "view stash" else "pickup stash"}§7.",
+            onClick = {
+                if (config.useViewStash) HypixelCommands.viewStash(lastType)
+                else HypixelCommands.pickupStash()
+            },
+            prefix = false,
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/StashCompact.kt]: Users will have a hard time finding this feature when they want to disable it. It's default enabled and has no prefix. Maybe we should still show the prefix?

**Review by hannibal002 — CHANGES_REQUESTED** [review]
The format doesn't look like SkyHanni to me. The colors are "off". Please use yellow/white/red ish colors
image

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**